File: src/pipeline/runner.ts
import path from 'node:path';
import { execFileSync } from 'node:child_process';
import fs from 'node:fs';
import {
  getResolutionPreset,
  getQualityPreset,
  AUTO_PRESET,
  type ResolutionPreset,
} from '../presets.js';
import { RenderError } from '../errors.js';
import {
  createTask,
  startTask,
  completeTask,
  failTask,
  updateProgress,
} from '../render/task.js';
import { normalizeClips } from '../render/transitions/shared.js';
import type { RenderResult } from '../render/types.js';
import type { QualityPreset } from '../presets.js';
import type {
  ResolvedPipelineClip,
  PipelineTransitionConfig,
} from './types.js';
import type { ParsedPipelineContext } from './config.js';

// ─── ffprobe helper ───────────────────────────────────────────────────────────

function probeSingleVideo(
  ffprobePath: string,
  videoPath: string,
): {
  duration: number;
  width: number;
  height: number;
  fps: number;
} {
  const raw = execFileSync(
    ffprobePath,
    [
      '-v', 'error', '-print_format', 'json',
      '-show_streams', '-show_format', '-select_streams', 'v:0',
      videoPath,
    ],
    { encoding: 'utf-8', timeout: 10000 },
  );
  const info = JSON.parse(raw) as {
    streams?: Array<{
      width?: number;
      height?: number;
      r_frame_rate?: string;
      duration?: string;
    }>;
    format?: { duration?: string };
  };
  const s = info.streams?.[0];
  const duration = Number(s?.duration ?? info.format?.duration ?? 0);
  const [num, den] = (s?.r_frame_rate ?? '30/1').split('/').map(Number);
  return {
    duration: Number.isFinite(duration) ? duration : 0,
    width: s?.width ?? 0,
    height: s?.height ?? 0,
    fps: den ? Math.round(num / den) : num || 30,
  };
}

// ─── FFmpeg filter_complex builder ────────────────────────────────────────────

/**
 * Generalised trim-mixed-concat with per-clip trim and per-junction transition duration.
 *
 * Input args layout: for each clip i:
 *   -ss {trimStart[i]}          (skipped if 0)
 *   -t  {effectiveDuration[i]}  (skipped if trimEnd is 0)
 *   -i  {clip[i].src}
 *
 * The filter_complex is structurally identical to ffmpegTrimMixedConcat
 * (src/render/transitions/trim-mixed-concat.ts) but uses:
 *   - effectiveDuration[i] per clip instead of a global (probedDuration - trimStart)
 *   - junctions[i].duration per junction instead of a shared transitionDuration
 */
function ffmpegPipelineConcat(
  ffmpegPath: string,
  clips: ResolvedPipelineClip[],
  junctions: PipelineTransitionConfig[],
  outputPath: string,
  qualPreset: QualityPreset,
  resPreset: ResolutionPreset,
  task: ReturnType<typeof createTask>,
): void {
  const n = clips.length;
  const fps = resPreset.fps;
  const frameDuration = 1 / fps;
  const minSeg = frameDuration / 2;
  const fmt = (v: number) => v.toFixed(6);

  // Per-clip -ss and -t input options
  const inputArgs: string[] = [];
  for (const clip of clips) {
    if (clip.trimStart > 0) inputArgs.push('-ss', String(clip.trimStart));
    if (clip.trimEnd > 0)   inputArgs.push('-t',  String(clip.effectiveDuration));
    inputArgs.push('-i', clip.src);
  }

  // Single-clip fast path
  if (n === 1) {
    execFileSync(
      ffmpegPath,
      [
        ...inputArgs,
        '-filter_complex', `[0:v]setpts=PTS-STARTPTS,fps=fps=${fps}[vout]`,
        '-map', '[vout]',
        '-c:v', 'libx264', '-preset', qualPreset.ffmpegPreset,
        '-crf', String(qualPreset.crf), '-pix_fmt', 'yuv420p', '-an',
        '-y', outputPath,
      ],
      { timeout: 600_000 },
    );
    updateProgress(task, 1);
    return;
  }

  const durations = clips.map((c) => c.effectiveDuration);

  // Compute per-clip head/tail consumed by bordering junctions
  const tailConsumed = new Array<number>(n).fill(0);
  const headConsumed = new Array<number>(n).fill(0);

  for (let j = 0; j < junctions.length; j++) {
    const { type: t, duration: D } = junctions[j];
    const halfD = D / 2;
    const dj = durations[j];
    const djn = durations[j + 1];

    if (t === 'flash-black') {
      const eff = Math.min(halfD, dj / 2, djn / 2);
      if (eff < halfD - 0.001) {
        console.warn(
          `  ⚠ 连接点 ${j + 1} (flash-black) 附近片段时长不足，闪黑已收缩至 ${(eff * 2).toFixed(2)}s`,
        );
      }
      tailConsumed[j] = eff;
      headConsumed[j + 1] = eff;
    } else if (t === 'dissolve') {
      const eff = Math.min(D, dj / 2);
      if (eff < D - 0.001) {
        console.warn(
          `  ⚠ 连接点 ${j + 1} (dissolve) 片段 ${clips[j].key} 时长不足，叠化已收缩至 ${eff.toFixed(2)}s`,
        );
      }
      tailConsumed[j] = eff;
      // dissolve does not consume next clip's head
    }
    // cut: no consumption
  }

  const filterParts: string[] = [];
  const segmentLabels: string[] = [];

  for (let i = 0; i < n; i++) {
    const d = durations[i];
    const bodyStart = headConsumed[i];
    const bodyDuration = d - headConsumed[i] - tailConsumed[i];

    if (bodyDuration > minSeg) {
      filterParts.push(
        `[${i}:v]trim=start=${fmt(bodyStart)}:duration=${fmt(bodyDuration)},setpts=PTS-STARTPTS,fps=fps=${fps}[body${i}]`,
      );
      segmentLabels.push(`[body${i}]`);
    } else if (bodyDuration > 0) {
      console.warn(`  ⚠ 片段 ${clips[i].key} 主体过短（${bodyDuration.toFixed(3)}s），已跳过`);
    }

    if (i < n - 1) {
      const { type: t, duration: D } = junctions[i];
      const halfD = D / 2;
      void halfD;

      if (t === 'flash-black') {
        const eff = tailConsumed[i];
        filterParts.push(
          `[${i}:v]trim=start=${fmt(d - eff)}:duration=${fmt(eff)},setpts=PTS-STARTPTS,` +
          `fade=t=out:st=0:d=${fmt(eff)},fps=fps=${fps}[fb_tail${i}]`,
        );
        segmentLabels.push(`[fb_tail${i}]`);
        filterParts.push(
          `[${i + 1}:v]trim=duration=${fmt(eff)},setpts=PTS-STARTPTS,` +
          `fade=t=in:st=0:d=${fmt(eff)},fps=fps=${fps}[fb_head${i + 1}]`,
        );
        segmentLabels.push(`[fb_head${i + 1}]`);
      } else if (t === 'dissolve') {
        const eff = tailConsumed[i];
        const stillDuration = Math.max(0, eff - frameDuration);
        filterParts.push(
          `[${i}:v]trim=start=${fmt(d - eff)}:duration=${fmt(eff)},setpts=PTS-STARTPTS,` +
          `format=rgba,fade=t=out:st=0:d=${fmt(eff)}:alpha=1,fps=fps=${fps}[diss_tail${i}]`,
        );
        filterParts.push(
          `[${i + 1}:v]trim=end_frame=1,setpts=PTS-STARTPTS,` +
          `tpad=stop_mode=clone:stop_duration=${fmt(stillDuration)},fps=fps=${fps}[diss_still${i}]`,
        );
        filterParts.push(
          `[diss_still${i}][diss_tail${i}]overlay=eof_action=pass:shortest=1,format=yuv420p,fps=fps=${fps}[diss_trans${i}]`,
        );
        segmentLabels.push(`[diss_trans${i}]`);
      }
      // cut: nothing added
    }
  }

  if (segmentLabels.length === 0) {
    throw new RenderError('没有可用的视频段（所有片段时长过短）');
  }

  filterParts.push(
    `${segmentLabels.join('')}concat=n=${segmentLabels.length}:v=1:a=0[vout]`,
  );

  const filterComplex = filterParts.join(';');
  const junctionDesc = junctions.map((j) => `${j.type}(${j.duration}s)`).join(' → ');
  console.log(`\n[2/3] FFmpeg Pipeline 拼接 ${n} 段（转场: ${junctionDesc}）...`);

  execFileSync(
    ffmpegPath,
    [
      ...inputArgs,
      '-filter_complex', filterComplex,
      '-map', '[vout]',
      '-c:v', 'libx264', '-preset', qualPreset.ffmpegPreset,
      '-crf', String(qualPreset.crf), '-pix_fmt', 'yuv420p', '-an',
      '-y', outputPath,
    ],
    { timeout: 600_000 },
  );
  updateProgress(task, 1);
}

// ─── PipelineRunner ───────────────────────────────────────────────────────────

export class PipelineRunner {
  private outputDir: string;

  constructor(private rootDir: string) {
    this.outputDir = path.join(rootDir, 'output');
  }

  async run(
    ctx: ParsedPipelineContext,
    ffmpegPath: string,
    ffprobePath: string,
    overrides: { preset?: string; quality?: string },
  ): Promise<RenderResult> {
    const { config, projectDir, configPath, resolvedSrcs, junctions } = ctx;

    const preset = overrides.preset ?? config.preset ?? AUTO_PRESET;
    const quality = overrides.quality ?? config.quality ?? 'high';
    const outputFilename = config.output?.filename ?? 'final.mp4';
    const projectName = path.basename(projectDir);

    const task = createTask('pipeline', {});
    startTask(task);

    const startTime = Date.now();
    const outDir = path.join(this.outputDir, projectName);
    fs.mkdirSync(outDir, { recursive: true });
    const outputPath = path.join(outDir, outputFilename);

    let cleanup: (() => void) | null = null;

    try {
      // ── 1. Probe each clip ──────────────────────────────────────────────
      console.log(`\n[1/3] 探测并校验 ${resolvedSrcs.length} 个片段...`);
      let resPreset: ResolutionPreset | null = null;

      const resolvedClips: ResolvedPipelineClip[] = resolvedSrcs.map((src, i) => {
        const clipCfg = config.clips[i];
        const trimStart = clipCfg.trim_start ?? 0;
        const trimEnd = clipCfg.trim_end ?? 0;
        const probed = probeSingleVideo(ffprobePath, src);
        const effectiveDuration = Math.max(0, probed.duration - trimStart - trimEnd);

        if (effectiveDuration <= 0) {
          throw new RenderError(
            `clips[${i}] (${path.basename(src)}): trim 后有效时长 ≤ 0，` +
            `探测到 ${probed.duration.toFixed(2)}s，trim_start=${trimStart}s，trim_end=${trimEnd}s`,
          );
        }
        console.log(
          `  clip_${i + 1}: ${path.basename(src)} ` +
          `(探测 ${probed.duration.toFixed(2)}s, 有效 ${effectiveDuration.toFixed(2)}s)`,
        );

        // Capture first clip's resolution for auto preset
        if (i === 0 && preset === AUTO_PRESET && probed.width && probed.height) {
          resPreset = {
            width: probed.width,
            height: probed.height,
            fps: probed.fps,
            label: `自动探测 ${probed.width}x${probed.height} ${probed.fps}fps`,
          };
          console.log(`  自动探测分辨率 ${probed.width}x${probed.height} ${probed.fps}fps`);
        }

        return {
          key: `clip_${i + 1}`,
          src,
          probedDuration: probed.duration,
          trimStart,
          trimEnd,
          effectiveDuration,
        };
      });

      // ── 2. Finalize resolution preset ───────────────────────────────────
      if (preset !== AUTO_PRESET) {
        resPreset = getResolutionPreset(preset);
      } else if (!resPreset) {
        console.warn('  ⚠ 无法探测分辨率，回退到 douyin_vertical (1080x1920)');
        resPreset = getResolutionPreset('douyin_vertical');
      }
      const qualPreset = getQualityPreset(quality);

      // ── 3. Normalize clips (scale/pad/fps) to temp files ────────────────
      // normalizeClips does NOT apply trim. Trimming happens in ffmpegPipelineConcat
      // via per-clip -ss / -t input options on the normalized temp files.
      const videoClips = resolvedClips.map((c) => ({
        key: c.key,
        src: c.src,
        duration: c.probedDuration,
      }));

      const normalized = normalizeClips(
        this.rootDir,
        ffmpegPath,
        videoClips,
        qualPreset,
        resPreset!,
      );
      cleanup = normalized.cleanup;

      // Re-attach trim params onto the normalized (temp) srcs
      const normalizedPipelineClips: ResolvedPipelineClip[] = normalized.clips.map(
        (nc, i) => ({ ...resolvedClips[i], src: nc.src }),
      );

      // ── 4. Build filter_complex and run FFmpeg ──────────────────────────
      ffmpegPipelineConcat(
        ffmpegPath,
        normalizedPipelineClips,
        junctions,
        outputPath,
        qualPreset,
        resPreset!,
        task,
      );

      // ── 5. Write meta and return ────────────────────────────────────────
      const elapsed = (Date.now() - startTime) / 1000;
      completeTask(task, outputPath);

      const meta = {
        taskId: task.id,
        mode: 'pipeline',
        preset,
        quality,
        configPath,
        clips: resolvedClips.map((c) => ({
          src: c.src,
          trimStart: c.trimStart,
          trimEnd: c.trimEnd,
          effectiveDuration: c.effectiveDuration,
        })),
        junctions,
        renderedAt: new Date().toISOString(),
        duration: Math.round(elapsed * 10) / 10,
        resolution: `${resPreset!.width}x${resPreset!.height}`,
      };
      fs.writeFileSync(
        path.join(outDir, 'meta.json'),
        JSON.stringify(meta, null, 2),
        'utf-8',
      );

      console.log(`\n[3/3] 归档产物...`);
      return { taskId: task.id, status: 'completed', outputPath, duration: elapsed };
    } catch (err) {
      const elapsed = (Date.now() - startTime) / 1000;
      const errMsg = err instanceof Error ? err.message : String(err);
      failTask(task, errMsg);
      return { taskId: task.id, status: 'failed', duration: elapsed, error: errMsg };
    } finally {
      if (cleanup) {
        try { cleanup(); } catch { /* ignore */ }
      }
    }
  }
}

File: videocut/pipeline/runner.py
from __future__ import annotations

import json
import random
import subprocess
import time
from datetime import datetime
from pathlib import Path

from videocut.bgm import apply_bgm, scan_bgm_files
from videocut.errors import RenderError
from videocut.ffmpeg_config import FFmpegVideoSettings, resolve_runtime_video_settings, resolve_video_settings
from videocut.log import get_logger
from videocut.pipeline.config import ParsedPipelineContext
from videocut.pipeline.types import PipelineTransitionConfig, ResolvedPipelineClip
from videocut.presets import AUTO_PRESET, QualityPreset, ResolutionPreset, get_quality_preset, get_resolution_preset
from videocut.render.task import complete_task, create_task, fail_task, start_task, update_progress
from videocut.render.transitions.shared import normalize_clips
from videocut.render.types import RenderResult, VideoClip


def probe_single_video(ffprobe_path: str, video_path: str) -> dict[str, float | int]:
    raw = subprocess.check_output(
        [
            ffprobe_path,
            "-v",
            "error",
            "-print_format",
            "json",
            "-show_streams",
            "-show_format",
            "-select_streams",
            "v:0",
            video_path,
        ],
        encoding="utf-8",
        timeout=10,
    )
    info = json.loads(raw)
    stream = (info.get("streams") or [None])[0] or {}
    duration = float(stream.get("duration") or info.get("format", {}).get("duration") or 0)
    num_str, den_str = str(stream.get("r_frame_rate") or "30/1").split("/")
    num = float(num_str)
    den = float(den_str)
    return {
        "duration": duration if duration == duration and duration != float("inf") else 0.0,
        "width": int(stream.get("width") or 0),
        "height": int(stream.get("height") or 0),
        "fps": int(round(num / den) if den else num or 30),
    }


def ffmpeg_pipeline_concat(
    ffmpeg_path: str,
    clips: list[ResolvedPipelineClip],
    junctions: list[PipelineTransitionConfig],
    output_path: str,
    qual_preset: QualityPreset,
    res_preset: ResolutionPreset,
    video_settings: FFmpegVideoSettings,
    task,
) -> None:
    clip_count = len(clips)
    fps = res_preset.fps
    frame_duration = 1 / fps
    min_segment = frame_duration / 2
    def _fmt(value: float) -> str:
        return f"{value:.6f}"

    input_args: list[str] = []
    for clip in clips:
        if clip.trim_start > 0:
            input_args.extend(["-ss", str(clip.trim_start)])
        if clip.trim_end > 0:
            input_args.extend(["-t", str(clip.effective_duration)])
        input_args.extend(["-i", clip.src])

    if clip_count == 1:
        subprocess.run(
            [
                ffmpeg_path,
                *video_settings.input_args(),
                *input_args,
                "-filter_complex",
                f"[0:v]setpts=PTS-STARTPTS,fps=fps={fps}[vout]",
                "-map",
                "[vout]",
                *video_settings.output_args(qual_preset),
                "-an",
                "-y",
                output_path,
            ],
            check=True,
            timeout=600,
        )
        update_progress(task, 1.0)
        return

    durations = [clip.effective_duration for clip in clips]
    tail_consumed = [0.0] * clip_count
    head_consumed = [0.0] * clip_count

    for index, junction in enumerate(junctions):
        current_duration = durations[index]
        next_duration = durations[index + 1]
        if junction.type == "flash-black":
            effective = min(junction.duration / 2, current_duration / 2, next_duration / 2)
            tail_consumed[index] = effective
            head_consumed[index + 1] = effective
        elif junction.type == "dissolve":
            tail_consumed[index] = min(junction.duration, current_duration / 2)

    filter_parts: list[str] = []
    segment_labels: list[str] = []
    for index in range(clip_count):
        duration = durations[index]
        body_start = head_consumed[index]
        body_duration = duration - head_consumed[index] - tail_consumed[index]
        if body_duration > min_segment:
            filter_parts.append(
                f"[{index}:v]trim=start={_fmt(body_start)}:duration={_fmt(body_duration)},"
                f"setpts=PTS-STARTPTS,fps=fps={fps}[body{index}]"
            )
            segment_labels.append(f"[body{index}]")

        if index >= clip_count - 1:
            continue
        junction = junctions[index]
        if junction.type == "flash-black":
            effective = tail_consumed[index]
            filter_parts.append(
                f"[{index}:v]trim=start={_fmt(duration - effective)}:duration={_fmt(effective)},"
                f"setpts=PTS-STARTPTS,fade=t=out:st=0:d={_fmt(effective)},fps=fps={fps}[fb_tail{index}]"
            )
            segment_labels.append(f"[fb_tail{index}]")
            filter_parts.append(
                f"[{index + 1}:v]trim=duration={_fmt(effective)},setpts=PTS-STARTPTS,"
                f"fade=t=in:st=0:d={_fmt(effective)},fps=fps={fps}[fb_head{index + 1}]"
            )
            segment_labels.append(f"[fb_head{index + 1}]")
        elif junction.type == "dissolve":
            effective = tail_consumed[index]
            still_duration = max(0.0, effective - frame_duration)
            filter_parts.append(
                f"[{index}:v]trim=start={_fmt(duration - effective)}:duration={_fmt(effective)},"
                f"setpts=PTS-STARTPTS,format=rgba,fade=t=out:st=0:d={_fmt(effective)}:alpha=1,"
                f"fps=fps={fps}[diss_tail{index}]"
            )
            filter_parts.append(
                f"[{index + 1}:v]trim=end_frame=1,setpts=PTS-STARTPTS,"
                f"tpad=stop_mode=clone:stop_duration={_fmt(still_duration)},fps=fps={fps}[diss_still{index}]"
            )
            filter_parts.append(
                f"[diss_still{index}][diss_tail{index}]overlay=eof_action=pass:shortest=1,"
                f"format=yuv420p,fps=fps={fps}[diss_trans{index}]"
            )
            segment_labels.append(f"[diss_trans{index}]")

    if not segment_labels:
        raise RenderError("No usable video segments after pipeline transition building.")

    filter_parts.append(f"{''.join(segment_labels)}concat=n={len(segment_labels)}:v=1:a=0[vout]")
    subprocess.run(
        [
            ffmpeg_path,
            *video_settings.input_args(),
            *input_args,
            "-filter_complex",
            ";".join(filter_parts),
            "-map",
            "[vout]",
            *video_settings.output_args(qual_preset),
            "-an",
            "-y",
            output_path,
        ],
        check=True,
        timeout=600,
    )
    update_progress(task, 1.0)


logger = get_logger(__name__)


class PipelineRunner:
    def __init__(self, root_dir: str | Path) -> None:
        self.root_dir = Path(root_dir).resolve()
        self.output_dir = self.root_dir / "output"

    def run(
        self,
        ctx: ParsedPipelineContext,
        ffmpeg_path: str,
        ffprobe_path: str,
        overrides: dict[str, str],
    ) -> RenderResult:
        config = ctx.config
        video_settings = resolve_runtime_video_settings(ffmpeg_path, resolve_video_settings())
        project_dir = ctx.project_dir
        config_path = ctx.config_path
        resolved_srcs = ctx.resolved_srcs
        junctions = ctx.junctions

        preset = overrides.get("preset") or config.preset or AUTO_PRESET
        quality = overrides.get("quality") or config.quality or "high"
        base_name = config.output.filename if config.output and config.output.filename else "final.mp4"
        stem, ext = Path(base_name).stem, Path(base_name).suffix or ".mp4"
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        output_filename = f"{stem}_{timestamp}{ext}"
        project_name = project_dir.name

        task = create_task("pipeline", {})
        start_task(task)

        start_time = time.time()
        out_dir = self.output_dir / project_name
        out_dir.mkdir(parents=True, exist_ok=True)
        output_path = str(out_dir / output_filename)
        cleanup = None

        try:
            logger.info("[1/3] Probing and validating %d clips...", len(resolved_srcs))
            res_preset = None
            resolved_clips: list[ResolvedPipelineClip] = []
            for index, src in enumerate(resolved_srcs):
                clip_cfg = config.clips[index]
                trim_start = clip_cfg.trim_start
                trim_end = clip_cfg.trim_end
                probed = probe_single_video(ffprobe_path, src)
                effective_duration = max(0.0, float(probed["duration"]) - trim_start - trim_end)
                if effective_duration <= 0:
                    raise RenderError(
                        f"clips[{index}] ({Path(src).name}) has non-positive duration after trim: "
                        f"source={probed['duration']:.2f}s trim_start={trim_start}s trim_end={trim_end}s"
                    )
                logger.info(
                    "  clip_%d: %s (source %.2fs, effective %.2fs)",
                    index + 1, Path(src).name, probed["duration"], effective_duration,
                )
                if index == 0 and preset == AUTO_PRESET and probed["width"] and probed["height"]:
                    res_preset = ResolutionPreset(
                        width=int(probed["width"]),
                        height=int(probed["height"]),
                        fps=int(probed["fps"]),
                        label=f"Auto {probed['width']}x{probed['height']} {probed['fps']}fps",
                    )
                    logger.info("  Auto-detected resolution %dx%d %dfps", probed["width"], probed["height"], probed["fps"])
                resolved_clips.append(
                    ResolvedPipelineClip(
                        key=f"clip_{index + 1}",
                        src=src,
                        probed_duration=float(probed["duration"]),
                        trim_start=trim_start,
                        trim_end=trim_end,
                        effective_duration=effective_duration,
                    )
                )

            if preset != AUTO_PRESET:
                res_preset = get_resolution_preset(preset)
            elif res_preset is None:
                logger.warning("unable to auto-detect resolution, fallback to douyin_vertical")
                res_preset = get_resolution_preset("douyin_vertical")

            qual_preset = get_quality_preset(quality)

            video_clips = [
                VideoClip(key=clip.key, src=clip.src, duration=clip.probed_duration)
                for clip in resolved_clips
            ]
            logger.info(
                "FFmpeg video encoder selected: %s%s",
                video_settings.encoder,
                f" (hwaccel={video_settings.hwaccel})" if video_settings.hwaccel else "",
            )
            normalized_clips, cleanup = normalize_clips(
                self.root_dir,
                ffmpeg_path,
                video_clips,
                qual_preset,
                res_preset,
                video_settings,
            )
            normalized_pipeline_clips = [
                ResolvedPipelineClip(
                    key=resolved_clips[index].key,
                    src=normalized_clips[index].src,
                    probed_duration=resolved_clips[index].probed_duration,
                    trim_start=resolved_clips[index].trim_start,
                    trim_end=resolved_clips[index].trim_end,
                    effective_duration=resolved_clips[index].effective_duration,
                )
                for index in range(len(resolved_clips))
            ]

            logger.info("[2/3] Building pipeline transitions...")
            ffmpeg_pipeline_concat(
                ffmpeg_path,
                normalized_pipeline_clips,
                junctions,
                output_path,
                qual_preset,
                res_preset,
                video_settings,
                task,
            )

            bgm_file_used = None
            if config.bgm and config.bgm.enabled:
                bgm_dir_path = (
                    Path(config.bgm.dir).resolve() if config.bgm.dir
                    else self.root_dir / "input" / "bgm"
                )
                bgm_files = scan_bgm_files(bgm_dir_path)
                chosen = random.choice(bgm_files)
                logger.info("[2.5/3] 混入 BGM: %s (volume=%.2f)", chosen.name, config.bgm.volume)
                apply_bgm(ffmpeg_path, ffprobe_path, output_path, chosen, config.bgm.volume, config.bgm.fade_out)
                bgm_file_used = str(chosen)

            elapsed = time.time() - start_time
            complete_task(task, output_path)
            meta = {
                "taskId": task.id,
                "mode": "pipeline",
                "preset": preset,
                "quality": quality,
                "configPath": str(config_path),
                "clips": [
                    {
                        "src": clip.src,
                        "trimStart": clip.trim_start,
                        "trimEnd": clip.trim_end,
                        "effectiveDuration": clip.effective_duration,
                    }
                    for clip in resolved_clips
                ],
                "junctions": [{"type": item.type, "duration": item.duration} for item in junctions],
                "renderedAt": datetime.utcnow().isoformat(),
                "duration": round(elapsed, 1),
                "resolution": f"{res_preset.width}x{res_preset.height}",
                "bgm": {
                    "file": bgm_file_used,
                    "volume": config.bgm.volume,
                    "fade_out": config.bgm.fade_out,
                } if bgm_file_used else None,
            }
            (out_dir / "meta.json").write_text(json.dumps(meta, indent=2, ensure_ascii=False), encoding="utf-8")
            logger.info("[3/3] Metadata written")
            return RenderResult(task_id=task.id, status="completed", output_path=output_path, duration=elapsed)
        except Exception as exc:  # intentional: isolate render failure from caller
            elapsed = time.time() - start_time
            fail_task(task, str(exc))
            return RenderResult(task_id=task.id, status="failed", duration=elapsed, error=str(exc))
        finally:
            if cleanup:
                try:
                    cleanup()
                except Exception:  # intentional: best-effort cleanup, never suppress render result
                    pass
